### ISSUE TITLE
no-loader-dupecheck can have ridiculous performance impact

### ISSUE THREAD

**Issue by solardiz** [openwall/john#5016]
At first I thought we had a rare infinite loop condition on processing of successful guesses, but no - the loop actually does terminate eventually. Here it took 47 seconds to remove 67k instances of one hash, and during this time the program was unresponsive:

```
$ rm tmp.pot; ./john -no-log -se=tmp -pot=tmp.pot -form=nt -mask=123456 -verb=1 -no-loader-dupecheck pp8-sample-10m
No dupe-checking performed when loading hashes.
Using default input encoding: UTF-8
Loaded 10000000 password hashes with no different salts (NT [MD4 128/128 AVX 4x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
66648g 0:00:00:47  1412g/s 0.02118p/s 0.02118c/s 211864C/s 123456
Session completed.
```

This is actually a reduced test case - I first saw the problem by trying to run actual cracking on a 100M file (not 10M above), and it appeared to get stuck right away for many minutes after loading, with no visible progress.

Somehow on a similar 1M file, the problem isn't seen - it's instant:

```
$ rm tmp.pot; ./john -no-log -se=tmp -pot=tmp.pot -form=nt -mask=123456 -verb=1 -no-loader-dupecheck pp8-sample-taken-1
No dupe-checking performed when loading hashes.
Using default input encoding: UTF-8
Loaded 1000000 password hashes with no different salts (NT [MD4 128/128 AVX 4x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
6617g 0:00:00:00  12253g/s 1.851p/s 1.851c/s 1851KC/s 123456
Session completed. 
```

Similarly, actual cracking on the 1M file just works, with no apparent slowdown.

I am puzzled by this, even though I wrote the code for removal of successful guesses.

**Comment by solardiz** [comment]
OK, not really puzzling: it's just O(n^2) when both the number of cracked passwords and the number of loaded hashes increase linearly with file size. So under 1 second for 1M, and 100x that for 10M, etc. Obviously, duplicates of a hash share the same hash bucket, and this turns the hash table lookup into a linear search.

Yet we could want to add this to description of the setting in `doc/OPTIONS` and `john.conf`.

**Comment by magnumripper** [comment]
I have seen pot syncs cause "hangs" for quite some time when very large numbers are removed in a sync. Should be the same problem.

BTW that's worse, because pot sync is the default.

**Comment by solardiz** [comment]
No, I think pot sync would be linear of the number of hashes to remove, unless duplicate hashes were loaded.
